gpytorch/lazy/matmul_lazy_variable.py

```diff
@@ -54,38 +65,48 @@ def _transpose_nonbatch(self, *args):
 
     def _batch_get_indices(self, batch_indices, left_indices, right_indices):
-        outer_size = batch_indices.size(0)
-        inner_size = self.lhs.size(-1)
-        inner_indices = Variable(right_indices.data.new(inner_size))
-        torch.arange(0, inner_size, out=inner_indices.data)
-
-        left_vals = self.lhs._batch_get_indices(
-            _outer_repeat(batch_indices, inner_size),
-            _outer_repeat(left_indices, inner_size),
-            _inner_repeat(inner_indices, outer_size),
-        )
-        right_vals = self.rhs._batch_get_indices(
-            _outer_repeat(batch_indices, inner_size),
-            _inner_repeat(inner_indices, outer_size),
-            _outer_repeat(right_indices, inner_size),
-        )
-
-        return (left_vals.view(-1, inner_size) * right_vals.view(-1, inner_size)).sum(-1)
+        n_indices = left_indices.numel()
+        if n_indices > self.size(-1) * self.size(-2) * self.size(-3):
+            return self._evaluated[batch_indices, left_indices, right_indices]
+
+        else:
+            outer_size = batch_indices.size(0)
+            inner_size = self.lhs.size(-1)
+            inner_indices = right_indices.new(inner_size)
+            torch.arange(0, inner_size, out=inner_indices.data)
+
+            left_vals = self.lhs._batch_get_indices(
+                _outer_repeat(batch_indices, inner_size),
+                _outer_repeat(left_indices, inner_size),
+                _inner_repeat(inner_indices, outer_size),
+            )
+            right_vals = self.rhs._batch_get_indices(
+                _outer_repeat(batch_indices, inner_size),
+                _inner_repeat(inner_indices, outer_size),
+                _outer_repeat(right_indices, inner_size),
+            )
+
+            return (left_vals.view(-1, inner_size) * right_vals.view(-1, inner_size)).sum(-1)
 
     def _get_indices(self, left_indices, right_indices):
-        outer_size = left_indices.size(0)
-        inner_size = self.lhs.size(-1)
-        inner_indices = Variable(right_indices.data.new(inner_size))
-        torch.arange(0, inner_size, out=inner_indices.data)
+        n_indices = left_indices.numel()
+        if n_indices > self.size(-1) * self.size(-2):
+            return self._evaluated[left_indices, right_indices]
+
+        else:
+            outer_size = left_indices.size(0)
+            inner_size = self.lhs.size(-1)
+            inner_indices = right_indices.new(inner_size)
+            torch.arange(0, inner_size, out=inner_indices.data)
 
-        left_vals = self.lhs._get_indices(
-            _outer_repeat(left_indices, inner_size), _inner_repeat(inner_indices, outer_size)
-        )
-        right_vals = self.rhs._get_indices(
-            _inner_repeat(inner_indices, outer_size), _outer_repeat(right_indices, inner_size)
-        )
+            left_vals = self.lhs._get_indices(
+                _outer_repeat(left_indices, inner_size), _inner_repeat(inner_indices, outer_size)
```

Comment:
nit: you're re-using `_inner_repeat(inner_indices, outer_size)` here, no need to compute it twice, just compute once and assign a variable

Reply:
Oops thanks for catching this!